File: app/src/main/java/com/sjn/stamp/controller/SongHistoryController.java
package com.sjn.stamp.controller;

import android.content.Context;
import android.support.v4.media.MediaMetadataCompat;

import com.sjn.stamp.constant.RecordType;
import com.sjn.stamp.db.Artist;
import com.sjn.stamp.db.Device;
import com.sjn.stamp.db.RankedArtist;
import com.sjn.stamp.db.RankedSong;
import com.sjn.stamp.db.Song;
import com.sjn.stamp.db.SongHistory;
import com.sjn.stamp.db.TotalSongHistory;
import com.sjn.stamp.db.dao.DeviceDao;
import com.sjn.stamp.db.dao.SongDao;
import com.sjn.stamp.db.dao.SongHistoryDao;
import com.sjn.stamp.db.dao.TotalSongHistoryDao;
import com.sjn.stamp.media.provider.ListProvider;
import com.sjn.stamp.ui.custom.TermSelectLayout;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.MediaRetrieveHelper;
import com.sjn.stamp.utils.NotificationHelper;
import com.sjn.stamp.utils.RealmHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.realm.Realm;

public class SongHistoryController {

    private static final String TAG = LogHelper.makeLogTag(SongHistoryController.class);

    private Context mContext;
    private DeviceDao mDeviceDao;
    private SongDao mSongDao;
    private SongHistoryDao mSongHistoryDao;
    private TotalSongHistoryDao mTotalSongHistoryDao;

    public SongHistoryController(Context context) {
        mContext = context;
        mDeviceDao = DeviceDao.getInstance();
        mSongDao = SongDao.getInstance();
        mSongHistoryDao = SongHistoryDao.getInstance();
        mTotalSongHistoryDao = TotalSongHistoryDao.getInstance();
    }

    public void deleteSongHistory(SongHistory songHistory) {
        Realm realm = RealmHelper.getRealmInstance();
        mSongHistoryDao.remove(realm, songHistory.getId());
        realm.close();
    }

    public void onPlay(MediaMetadataCompat track, Date date) {
        LogHelper.i(TAG, "insertPLAY ", track.getDescription().getTitle());
        registerHistory(track, RecordType.PLAY, date);
    }

    public void onSkip(MediaMetadataCompat track, Date date) {
        LogHelper.i(TAG, "insertSKIP ", track.getDescription().getTitle());
        registerHistory(track, RecordType.SKIP, date);
    }

    public void onStart(MediaMetadataCompat track, Date date) {
        LogHelper.i(TAG, "insertSTART ", track.getDescription().getTitle());
        registerHistory(track, RecordType.START, date);
    }

    public void onComplete(MediaMetadataCompat track, Date date) {
        LogHelper.i(TAG, "insertComplete ", track.getDescription().getTitle());
        registerHistory(track, RecordType.COMPLETE, date);
    }

    private void registerHistory(MediaMetadataCompat track, RecordType recordType, Date date) {
        Song song = createSong(track);
        Realm realm = RealmHelper.getRealmInstance();
        int playCount = mTotalSongHistoryDao.saveOrIncrement(realm, createTotalSongHistory(song, recordType));
        mSongHistoryDao.save(realm, createSongHistory(song, createDevice(), recordType, date, playCount));
        if (recordType == RecordType.PLAY && isSendNotification(playCount)) {
            SongHistory oldestSongHistory = mSongHistoryDao.findOldest(realm, song);
            NotificationHelper.sendNotification(mContext, song.getTitle(), playCount, oldestSongHistory.mRecordedAt);
        }
        realm.close();
    }

    private boolean isSendNotification(int count) {
        return count == 10 || count == 50 || (count % 100 == 0 && count >= 100);
    }

    private TotalSongHistory createTotalSongHistory(Song song, RecordType recordType) {
        TotalSongHistory totalSongHistory = mTotalSongHistoryDao.newStandalone();
        totalSongHistory.parseSongQueue(song, recordType);
        return totalSongHistory;
    }

    private Device createDevice() {
        Device device = mDeviceDao.newStandalone();
        device.configure();
        return device;
    }

    private Song createSong(MediaMetadataCompat track) {
        Song song = mSongDao.newStandalone();
        song.parseMetadata(track);
        return song;
    }

    private SongHistory createSongHistory(Song song, Device device, RecordType recordType, Date date, int count) {
        SongHistory songHistory = mSongHistoryDao.newStandalone();
        songHistory.setValues(song, recordType, device, date, count);
        return songHistory;
    }

    public List<MediaMetadataCompat> getTopSongList() {
        Realm realm = RealmHelper.getRealmInstance();
        List<MediaMetadataCompat> trackList = new ArrayList<>();
        List<TotalSongHistory> historyList = mTotalSongHistoryDao.getOrderedList(realm);
        for (TotalSongHistory totalSongHistory : historyList) {
            if (totalSongHistory.getPlayCount() == 0 || trackList.size() > 30) {
                break;
            }
            //noinspection ResourceType
            trackList.add(
                    totalSongHistory.getSong().mediaMetadataCompatBuilder()
                            .putLong(ListProvider.CUSTOM_METADATA_TRACK_PREFIX, totalSongHistory.getPlayCount())
                            .build()
            );
        }
        realm.close();
        return trackList;
    }

    public List<SongHistory> getManagedTimeline(Realm realm) {
        return mSongHistoryDao.timeline(realm, RecordType.PLAY.getValue());
    }

    public List<RankedSong> getRankedSongList(Realm realm, TermSelectLayout.Term term) {
        return getRankedSongList(realm,
                term.from() == null ? null : term.from().toDateTimeAtStartOfDay().toDate(),
                term.to() == null ? null : term.to().toDateTimeAtStartOfDay().plusDays(1).toDate());
    }

    public List<RankedArtist> getRankedArtistList(TermSelectLayout.Term term) {
        return getRankedArtistList(
                term.from() == null ? null : term.from().toDateTimeAtStartOfDay().toDate(),
                term.to() == null ? null : term.to().toDateTimeAtStartOfDay().plusDays(1).toDate());
    }

    private List<RankedSong> getRankedSongList(Realm realm, Date from, Date to) {
        LogHelper.i(TAG, "getRankedSongList start");
        LogHelper.i(TAG, "calc historyList");
        List<SongHistory> historyList = mSongHistoryDao.where(realm, from, to, RecordType.PLAY.getValue());
        Map<Song, Integer> songCountMap = new HashMap<>();
        LogHelper.i(TAG, "put songCountMap");
        for (SongHistory songHistory : historyList) {
            if (songCountMap.containsKey(songHistory.getSong())) {
                songCountMap.put(songHistory.getSong(), songCountMap.get(songHistory.getSong()) + 1);
            } else {
                songCountMap.put(songHistory.getSong(), 0);
            }
        }
        LogHelper.i(TAG, "create rankedSongList");
        List<RankedSong> rankedSongList = new ArrayList<>();
        for (Map.Entry<Song, Integer> entry : songCountMap.entrySet()) {
            rankedSongList.add(new RankedSong(entry.getValue(), entry.getKey()));
        }
        LogHelper.i(TAG, "sort rankedSongList");
        Collections.sort(rankedSongList, new Comparator<RankedSong>() {
            @Override
            public int compare(RankedSong t1, RankedSong t2) {
                return t2.getPlayCount() - t1.getPlayCount();
            }
        });
        if (rankedSongList.size() > 30) {
            rankedSongList = rankedSongList.subList(0, 30);
        }
        LogHelper.i(TAG, "getRankedSongList end");
        return rankedSongList;
    }

    private List<RankedArtist> getRankedArtistList(Date from, Date to) {
        LogHelper.i(TAG, "getRankedArtistList start");
        Realm realm = RealmHelper.getRealmInstance();
        List<SongHistory> historyList = mSongHistoryDao.where(realm, from, to, RecordType.PLAY.getValue());
        Map<String, Integer> artistMap = new HashMap<>();
        for (SongHistory songHistory : historyList) {
            String artist = songHistory.getSong().getArtist();
            int count = artistMap.containsKey(artist) ? artistMap.get(artist) + 1 : 1;
            artistMap.put(artist, count);
        }
        realm.close();
        List<RankedArtist> rankedArtistList = new ArrayList<>();
        for (Map.Entry<String, Integer> e : artistMap.entrySet()) {
            rankedArtistList.add(new RankedArtist(e.getValue(), new Artist(e.getKey(), MediaRetrieveHelper.findAlbumArtByArtist(mContext, e.getKey(), new MediaRetrieveHelper.PermissionRequiredCallback() {
                @Override
                public void onPermissionRequired() {

                }
            }))));
        }
        Collections.sort(rankedArtistList, new Comparator<RankedArtist>() {
            @Override
            public int compare(RankedArtist t1, RankedArtist t2) {
                return t2.getPlayCount() - t1.getPlayCount();
            }
        });
        LogHelper.i(TAG, "getRankedArtistList end");
        return rankedArtistList;
    }
}

File: app/src/main/java/com/sjn/stamp/controller/SongHistoryController.kt
package com.sjn.stamp.controller

import android.content.Context
import android.os.AsyncTask
import android.support.v4.media.MediaMetadataCompat
import com.sjn.stamp.model.*
import com.sjn.stamp.model.constant.RecordType
import com.sjn.stamp.model.dao.SongDao
import com.sjn.stamp.model.dao.SongHistoryDao
import com.sjn.stamp.model.dao.TotalSongHistoryDao
import com.sjn.stamp.ui.custom.PeriodSelectLayout
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.MediaIDHelper.resolveMusicId
import com.sjn.stamp.utils.MediaItemHelper
import com.sjn.stamp.utils.NotificationHelper
import com.sjn.stamp.utils.RealmHelper
import io.realm.Realm
import java.lang.ref.WeakReference
import java.util.*

class SongHistoryController(private val context: Context) {

    val topSongList: List<MediaMetadataCompat>
        get() =
             RealmHelper.realmInstance.use { realm ->
                val songList = ArrayList<MediaMetadataCompat>()
                TotalSongHistoryDao.findPlayed(realm).forEach loop@ { songHistory ->
                    if (songHistory.playCount == 0 || songList.size >= UserSettingController().mostPlayedSongSize) {
                        return@loop
                    }
                    songHistory.song?.let {
                        if (it.isNotEmpty()) {
                            songList.add(MediaItemHelper.convertToMetadata(songHistory.song.first()))
                        }
                    }
                }
                songList
            }

    val hasHistory: Boolean
        get() =
            RealmHelper.realmInstance.use { realm ->
                SongHistoryDao.isExists(realm)
            }

    fun onPlay(mediaId: String, date: Date) {
        LogHelper.d(TAG, "insertPLAY ", mediaId)
        register(mediaId, RecordType.PLAY, date)
    }

    fun onSkip(mediaId: String, date: Date) {
        LogHelper.d(TAG, "insertSKIP ", mediaId)
        register(mediaId, RecordType.SKIP, date)
    }

    fun onStart(mediaId: String, date: Date) {
        LogHelper.d(TAG, "insertSTART ", mediaId)
        register(mediaId, RecordType.START, date)
    }

    fun onComplete(mediaId: String, date: Date) {
        LogHelper.d(TAG, "insertComplete ", mediaId)
        register(mediaId, RecordType.COMPLETE, date)
    }

    fun delete(songHistoryId: Long) {
        RealmHelper.realmInstance.use { realm ->
            SongHistoryDao.delete(realm, songHistoryId)
        }
    }

    fun getManagedTimeline(realm: Realm): List<SongHistory> = SongHistoryDao.timeline(realm, RecordType.PLAY.databaseValue)

    fun getRankedSongList(realm: Realm, period: PeriodSelectLayout.Period): List<RankedSong> = getRankedSongList(realm, period.from(), period.to(), 30)

    fun getRankedArtistList(realm: Realm, period: PeriodSelectLayout.Period): List<RankedArtist> = getRankedArtistList(realm, period.from(), period.to(), 30)

    private fun register(mediaId: String, recordType: RecordType, date: Date) {
        val musicId = resolveMusicId(mediaId) ?: return
        RealmHelper.realmInstance.use { realm ->
            val song = SongDao.findOrCreateByMusicId(realm, musicId, context)
            song ?: return
            val playCount = TotalSongHistoryDao.increment(realm, song.id, recordType)
            SongHistoryDao.create(realm, song, recordType, date, playCount)
            if (recordType === RecordType.PLAY) {
                sendNotificationBySongCount(realm, song, playCount)
                sendNotificationByArtistCount(song)
            }
            SongController(context).calculateSmartStamp(song.id, playCount, recordType)
        }
    }

    private fun sendNotificationBySongCount(realm: Realm, song: Song, playCount: Int) {
        if (NotificationHelper.isSendPlayedNotification(playCount)) {
            NotificationHelper.sendPlayedNotification(context, song, song.albumArtUri, playCount, SongHistoryDao.findOldest(realm, song.id)!!.recordedAt)
        }
    }

    private fun sendNotificationByArtistCount(song: Song) {
        ArtistCountAsyncTask(context, song.artist.name).execute()
    }

    private fun getRankedSongList(realm: Realm, from: Date?, to: Date?, count: Int?): List<RankedSong> {
        LogHelper.d(TAG, "getRankedSongList start")
        LogHelper.d(TAG, "calc historyList")
        val songCountMap = HashMap<Song, Int>()
        LogHelper.d(TAG, "put songCountMap")
        for (songHistory in SongHistoryDao.where(realm, from, to, RecordType.PLAY.databaseValue)) {
            if (songCountMap.containsKey(songHistory.song)) {
                songCountMap[songHistory.song] = songCountMap[songHistory.song]!! + 1
            } else {
                songCountMap[songHistory.song] = 1
            }
        }
        LogHelper.d(TAG, "create rankedSongList")
        var rankedSongList: MutableList<RankedSong> = ArrayList()
        for ((key, value) in songCountMap) {
            rankedSongList.add(RankedSong(value, key))
        }
        LogHelper.d(TAG, "sort rankedSongList")
        rankedSongList.sortWith(Comparator { t1, t2 -> t2.playCount - t1.playCount })
        count?.let {
            if (rankedSongList.size > count) {
                rankedSongList = rankedSongList.subList(0, count)
            }
        }
        LogHelper.d(TAG, "getRankedSongList end")
        return rankedSongList
    }

    fun getRankedArtistList(realm: Realm, from: Date?, to: Date?, count: Int?): List<RankedArtist> {
        LogHelper.d(TAG, "getRankedArtistList start")
        val historyList = SongHistoryDao.where(realm, from, to, RecordType.PLAY.databaseValue)
        val artistMap = HashMap<Artist, ArtistCounter>()
        for (songHistory in historyList) {
            ArtistCounter.count(artistMap, songHistory.song.artist, songHistory.song)
        }
        var rankedArtistList: MutableList<RankedArtist> = ArrayList()
        for ((key, value) in artistMap) {
            rankedArtistList.add(RankedArtist(value.count, key, value.songCountMap))
        }
        rankedArtistList.sortWith(Comparator { t1, t2 -> t2.playCount - t1.playCount })
        count?.let {
            if (rankedArtistList.size > count) {
                rankedArtistList = rankedArtistList.subList(0, count)
            }
        }
        LogHelper.d(TAG, "getRankedArtistList end")
        return rankedArtistList
    }

    private class ArtistCounter {
        internal var count = 0
        internal var songCountMap: MutableMap<Song, Int> = HashMap()

        internal fun increment(song: Song) {
            count++
            if (songCountMap.containsKey(song)) {
                songCountMap[song] = songCountMap[song]!! + 1
            } else {
                songCountMap[song] = 1
            }
        }

        companion object {

            fun count(artistMap: MutableMap<Artist, ArtistCounter>, artist: Artist, song: Song) {
                val counter = if (artistMap.containsKey(artist)) artistMap[artist] else ArtistCounter()
                counter?.let { artistCounter ->
                    artistCounter.increment(song)
                    artistMap.put(artist, artistCounter)
                }

            }
        }
    }

    companion object {

        private val TAG = LogHelper.makeLogTag(SongHistoryController::class.java)

        private class ArtistCountAsyncTask constructor(context: Context, internal var artistName: String) : AsyncTask<Void, Void, Void>() {

            val context = WeakReference(context)
            override fun doInBackground(vararg params: Void): Void? {
                RealmHelper.realmInstance.use { realm ->
                    val historyList = SongHistoryDao.findPlayRecordByArtist(realm, artistName)
                    if (!NotificationHelper.isSendPlayedNotification(historyList.size)) {
                        return null
                    }
                    val oldestSongHistory = SongHistoryDao.findOldestByArtist(realm, artistName)
                    context.get()?.let {
                        NotificationHelper.sendPlayedNotification(it, artistName, oldestSongHistory!!.song.albumArtUri, historyList.size, oldestSongHistory.recordedAt)
                    }
                }
                return null
            }
        }
    }

}
